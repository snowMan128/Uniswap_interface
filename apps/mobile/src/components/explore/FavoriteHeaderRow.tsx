import { default as React } from 'react'
import { useTranslation } from 'react-i18next'
import { Flex, Text, TouchableArea } from 'ui/src'
import { TripleDots } from 'ui/src/components/icons'
import { iconSizes } from 'ui/src/theme'

export function FavoriteHeaderRow({
  title,
  editingTitle,
  isEditing,
  onPress,
}: {
  title: string
  editingTitle: string
  isEditing: boolean
  onPress: () => void
}): JSX.Element {
  const { t } = useTranslation()
  return (
    <Flex
      row
      alignItems="center"
      gap="$spacing16"
      justifyContent="space-between"
      mb="$spacing8"
      mx="$spacing8">
      <Text color="$neutral2" variant="subheading2">
        {isEditing ? editingTitle : title}
      </Text>
      {!isEditing ? (
        <TouchableArea
          hapticFeedback
          hitSlop={16}
          testID="favorite-header-row/favorite-button"
          onPress={onPress}>
          <TripleDots
            color="$neutral2"
            size={iconSizes.icon20}
            strokeLinecap="round"
            strokeWidth={1}
          />
        </TouchableArea>
      ) : (
        <TouchableArea hitSlop={16} onPress={onPress}>
          <Text color="$accent1" testID="favorite-header-row/done-button" variant="buttonLabel3">
            {t('common.button.done')}
          </Text>
        </TouchableArea>
      )}
    </Flex>
  )
}
