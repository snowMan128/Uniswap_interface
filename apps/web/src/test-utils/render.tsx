import { MockedProvider } from '@apollo/client/testing'
import { queries } from '@testing-library/dom'
import { render, renderHook, RenderHookOptions, RenderOptions } from '@testing-library/react'
import { AssetActivityProvider } from 'graphql/data/apollo/AssetActivityProvider'
import { TokenBalancesProvider } from 'graphql/data/apollo/TokenBalancesProvider'
import { BlockNumberContext } from 'lib/hooks/useBlockNumber'
import { PropsWithChildren, ReactElement, ReactNode } from 'react'
import { HelmetProvider } from 'react-helmet-async/lib/index'
import { QueryClient, QueryClientProvider } from 'react-query'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import store from 'state'
import { ThemeProvider } from 'theme'
import { TamaguiProvider } from 'theme/tamaguiProvider'
import { UnitagUpdaterContextProvider } from 'uniswap/src/features/unitags/context'

const queryClient = new QueryClient()

const BLOCK_NUMBER_CONTEXT = { fastForward: () => {}, block: 1234, mainnetBlock: 1234 }
function MockedBlockNumberProvider({ children }: PropsWithChildren) {
  return <BlockNumberContext.Provider value={BLOCK_NUMBER_CONTEXT}>{children}</BlockNumberContext.Provider>
}

const WithProviders = ({ children }: { children?: ReactNode }) => {
  return (
    <HelmetProvider>
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          <BrowserRouter>
            {/*
             * Web3Provider is mocked through setupTests.ts
             * To test behavior that depends on Web3Provider, use jest.unmock('@web3-react/core')
             */}
            <MockedProvider showWarnings={false}>
              <AssetActivityProvider>
                <TokenBalancesProvider>
                  <MockedBlockNumberProvider>
                    <UnitagUpdaterContextProvider>
                      <ThemeProvider>
                        <TamaguiProvider>{children}</TamaguiProvider>
                      </ThemeProvider>
                    </UnitagUpdaterContextProvider>
                  </MockedBlockNumberProvider>
                </TokenBalancesProvider>
              </AssetActivityProvider>
            </MockedProvider>
          </BrowserRouter>
        </QueryClientProvider>
      </Provider>
    </HelmetProvider>
  )
}

type CustomRenderOptions = Omit<RenderOptions, 'wrapper'>
const customRender = (ui: ReactElement, options?: CustomRenderOptions) => {
  return render<typeof queries>(ui, { ...options, wrapper: WithProviders })
}

type CustomRenderHookOptions<Props> = Omit<RenderHookOptions<Props>, 'wrapper'>
const customRenderHook = <Result, Props>(
  hook: (initialProps: Props) => Result,
  options?: CustomRenderHookOptions<Props>
) => {
  return renderHook(hook, { ...options, wrapper: WithProviders })
}

// Testing utils may export *.
// eslint-disable-next-line no-restricted-syntax
export * from '@testing-library/react'
export { customRender as render, customRenderHook as renderHook }
