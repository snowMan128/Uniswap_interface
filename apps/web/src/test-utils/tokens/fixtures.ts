import { QueryResult } from '@apollo/client'
import { Chain, Exact, TokenProjectQuery } from 'graphql/data/__generated__/types-and-hooks'

export const validTokenProjectResponse = {
  data: {
    token: {
      id: 'VG9rZW46RVRIRVJFVU1fMHhBMGI4Njk5MWM2MjE4YjM2YzFkMTlENGEyZTlFYjBjRTM2MDZlQjQ4',
      decimals: 6,
      name: 'USD Coin',
      chain: 'ETHEREUM',
      address: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
      symbol: 'USDC',
      standard: 'ERC20',
      project: {
        id: 'VG9rZW5Qcm9qZWN0OkVUSEVSRVVNXzB4YTBiODY5OTFjNjIxOGIzNmMxZDE5ZDRhMmU5ZWIwY2UzNjA2ZWI0OF9VU0RD',
        description:
          'USDC is a fully collateralized US dollar stablecoin. USDC is the bridge between dollars and trading on cryptocurrency exchanges. The technology behind CENTRE makes it possible to exchange value between people, businesses and financial institutions just like email between mail services and texts between SMS providers. We believe by removing artificial economic borders, we can create a more inclusive global economy.',
        homepageUrl: 'https://www.circle.com/en/usdc',
        twitterName: 'circle',
        logoUrl:
          'https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/logo.png',
        __typename: 'TokenProject',
      },
      __typename: 'Token',
    },
  },
} as unknown as QueryResult<TokenProjectQuery, Exact<{ chain: Chain; address?: string }>>
