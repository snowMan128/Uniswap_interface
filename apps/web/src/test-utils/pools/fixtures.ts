import { BigNumber } from '@ethersproject/bignumber'
import { ChainId, Currency, WETH9 } from '@uniswap/sdk-core'
import { FeeAmount, Pool, Position } from '@uniswap/v3-sdk'
import { USDC_MAINNET } from 'constants/tokens'
import { Token as BEToken } from 'graphql/data/__generated__/types-and-hooks'
import { PoolData } from 'graphql/data/pools/usePoolData'
import { Token } from 'graphql/thegraph/__generated__/types-and-hooks'

export const validParams = { poolAddress: '0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640', chainName: 'ethereum' }

export const validPoolToken0 = {
  id: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
  symbol: 'USDC',
  name: 'USD Coin',
  decimals: '6',
  derivedETH: '0.0006240873011635544626425964678706127',
  __typename: 'Token',
} as Token
export const validBEPoolToken0 = {
  ...validPoolToken0,
  address: validPoolToken0.id,
  chain: 'ETHEREUM',
  decimals: 6,
} as BEToken

export const validUSDCCurrency = {
  isNative: false,
  isToken: true,
  name: 'USDCoin',
  address: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
  symbol: 'USDC',
  decimals: 6,
  chainId: 1,
  logoURI:
    'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/logo.png',
  _checksummedAddress: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
  _tags: null,
  wrapped: validPoolToken0,
} as unknown as Currency

const validPoolToken1 = {
  id: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
  symbol: 'WETH',
  name: 'Wrapped Ether',
  decimals: '18',
  derivedETH: '1',
  __typename: 'Token',
} as Token
export const validBEPoolToken1 = {
  ...validPoolToken1,
  address: validPoolToken1.id,
  chain: 'ETHEREUM',
  decimals: 18,
} as BEToken

export const owner = '0xf5b6bb25f5beaea03dd014c6ef9fa9f3926bf36c'

const pool = new Pool(
  USDC_MAINNET,
  WETH9[ChainId.MAINNET],
  FeeAmount.MEDIUM,
  '1851127709498178402383049949138810',
  '7076437181775065414',
  201189
)

const position = new Position({
  pool,
  liquidity: 1341008833950736,
  tickLower: 200040,
  tickUpper: 202560,
})
const details = {
  nonce: BigNumber.from('0'),
  tokenId: BigNumber.from('0'),
  operator: '0x0',
  token0: USDC_MAINNET.address,
  token1: WETH9[ChainId.MAINNET].address,
  fee: FeeAmount.MEDIUM,
  tickLower: -100,
  tickUpper: 100,
  liquidity: BigNumber.from('9000'),
  feeGrowthInside0LastX128: BigNumber.from('0'),
  feeGrowthInside1LastX128: BigNumber.from('0'),
  tokensOwed0: BigNumber.from('0'),
  tokensOwed1: BigNumber.from('0'),
}
export const useMultiChainPositionsReturnValue = {
  positions: [
    {
      owner,
      chainId: ChainId.MAINNET,
      position,
      pool,
      details,
      inRange: true,
      closed: false,
    },
  ],
  loading: false,
}

export const usdcWethPoolAddress = '0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640'

export const validPoolDataResponse = {
  data: {
    feeTier: 500,
    token0: validBEPoolToken0,
    token1: validBEPoolToken1,
    token0Price: 1605.481,
    token1Price: 0.000622,
    txCount: 5456494,
    address: usdcWethPoolAddress,
    volumeUSD24HChange: -17.753809465717136,
    volumeUSD24H: 194273059.265625,
    tvlUSD: 223166198.4690675,
    tvlUSDChange: -0.3657085465786977,
    tvlToken0: 90930713.7356909,
    tvlToken1: 82526.48678530742,
  } as PoolData,
  loading: false,
  error: false,
}
