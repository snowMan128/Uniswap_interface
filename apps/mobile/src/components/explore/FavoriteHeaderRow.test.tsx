import { fireEvent, render } from 'src/test/test-utils'
import { ON_PRESS_EVENT_PAYLOAD } from 'wallet/src/test/fixtures'
import { FavoriteHeaderRow } from './FavoriteHeaderRow'

const defaultProps = {
  title: 'Title',
  editingTitle: 'Editing Title',
  isEditing: false,
  onPress: jest.fn(),
}

describe(FavoriteHeaderRow, () => {
  describe('when not editing', () => {
    it('renders without error', () => {
      const tree = render(<FavoriteHeaderRow {...defaultProps} />)

      expect(tree.toJSON()).toMatchSnapshot()
    })

    it('renders title', () => {
      const { queryByText } = render(<FavoriteHeaderRow {...defaultProps} />)

      expect(queryByText(defaultProps.title)).toBeTruthy()
      expect(queryByText(defaultProps.editingTitle)).toBeFalsy()
    })

    it('renders favorite button', () => {
      const { queryByTestId } = render(<FavoriteHeaderRow {...defaultProps} />)

      const favoriteButton = queryByTestId('favorite-header-row/favorite-button')
      const doneButton = queryByTestId('favorite-header-row/done-button')

      expect(favoriteButton).toBeTruthy()
      expect(doneButton).toBeFalsy()
    })

    it('calls onPress when favorite icon pressed', () => {
      const { getByTestId } = render(<FavoriteHeaderRow {...defaultProps} />)

      const favoriteButton = getByTestId('favorite-header-row/favorite-button')
      fireEvent.press(favoriteButton, ON_PRESS_EVENT_PAYLOAD)

      expect(defaultProps.onPress).toHaveBeenCalledTimes(1)
    })
  })

  describe('when editing', () => {
    it('renders without error', () => {
      const tree = render(<FavoriteHeaderRow {...defaultProps} isEditing />)

      expect(tree.toJSON()).toMatchSnapshot()
    })

    it('renders editingTitle', () => {
      const { queryByText } = render(<FavoriteHeaderRow {...defaultProps} isEditing />)

      expect(queryByText(defaultProps.editingTitle)).toBeTruthy()
      expect(queryByText(defaultProps.title)).toBeFalsy()
    })

    it('renders done button', () => {
      const { queryByTestId } = render(<FavoriteHeaderRow {...defaultProps} isEditing />)

      const favoriteButton = queryByTestId('favorite-header-row/favorite-button')
      const doneButton = queryByTestId('favorite-header-row/done-button')

      expect(favoriteButton).toBeFalsy()
      expect(doneButton).toBeTruthy()
    })

    it('calls onPress when done button pressed', () => {
      const { getByTestId } = render(<FavoriteHeaderRow {...defaultProps} isEditing />)

      const doneButton = getByTestId('favorite-header-row/done-button')
      fireEvent.press(doneButton, ON_PRESS_EVENT_PAYLOAD)

      expect(defaultProps.onPress).toHaveBeenCalledTimes(1)
    })
  })
})
