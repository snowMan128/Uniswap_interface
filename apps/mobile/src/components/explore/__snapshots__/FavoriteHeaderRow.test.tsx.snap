// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteHeaderRow when editing renders without error 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 16,
      "justifyContent": "space-between",
      "marginBottom": 8,
      "marginLeft": 8,
      "marginRight": 8,
    }
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    style={
      {
        "color": "#7D7D7D",
        "fontFamily": "Basel-Book",
        "fontSize": 17,
        "lineHeight": 24,
      }
    }
    suppressHighlighting={true}
  >
    Editing Title
  </Text>
  <View
    cancelable={true}
    disabled={false}
    focusable={true}
    hitSlop={[Function]}
    minPressDuration={0}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onPress={[Function]}
    onPressIn={[Function]}
    onPressOut={[Function]}
    style={
      {
        "flexDirection": "column",
        "opacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
  >
    <Text
      allowFontScaling={true}
      maxFontSizeMultiplier={1.2}
      style={
        {
          "color": "#FC72FF",
          "fontFamily": "Basel-Medium",
          "fontSize": 17,
          "fontWeight": "500",
          "lineHeight": 24,
        }
      }
      suppressHighlighting={true}
      testID="favorite-header-row/done-button"
    >
      Done
    </Text>
  </View>
</View>
`;

exports[`FavoriteHeaderRow when not editing renders without error 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 16,
      "justifyContent": "space-between",
      "marginBottom": 8,
      "marginLeft": 8,
      "marginRight": 8,
    }
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    style={
      {
        "color": "#7D7D7D",
        "fontFamily": "Basel-Book",
        "fontSize": 17,
        "lineHeight": 24,
      }
    }
    suppressHighlighting={true}
  >
    Title
  </Text>
  <View
    cancelable={true}
    disabled={false}
    focusable={true}
    hitSlop={[Function]}
    minPressDuration={0}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onPress={[Function]}
    onPressIn={[Function]}
    onPressOut={[Function]}
    style={
      {
        "flexDirection": "column",
        "opacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
    testID="favorite-header-row/favorite-button"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="20"
      bbWidth="20"
      fill="currentColor"
      focusable={false}
      meetOrSlice={0}
      minX={0}
      minY={0}
      stroke="currentColor"
      strokeLinecap="round"
      strokeWidth={1}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "color": "#7D7D7D",
            "height": 20,
            "width": 20,
          },
          {
            "flex": 0,
            "height": 20,
            "width": 20,
          },
        ]
      }
      tintColor="#7D7D7D"
      vbHeight={4}
      vbWidth={18}
    >
      <RNSVGGroup
        fill={
          {
            "type": 2,
          }
        }
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeWidth="1"
      >
        <RNSVGPath
          d="M9 3a1 1 0 1 0 0-2 1 1 0 0 0 0 2Z"
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="2"
        />
        <RNSVGPath
          d="M16 3a1 1 0 1 0 0-2 1 1 0 0 0 0 2Z"
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="2"
        />
        <RNSVGPath
          d="M2 3a1 1 0 1 0 0-2 1 1 0 0 0 0 2Z"
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="2"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
